src/gs-installed-page.c

```diff
@@ -546,44 +553,74 @@ gs_installed_page_has_app (GsInstalledPage *self,
 }
 
 static void
-gs_installed_page_pending_apps_changed_cb (GsPluginLoader *plugin_loader,
-                                           GsInstalledPage *self)
+gs_installed_page_pending_apps_refined_cb (GObject *source,
+					   GAsyncResult *res,
+					   gpointer user_data)
 {
-	GsApp *app;
+	GsPluginLoader *plugin_loader = GS_PLUGIN_LOADER (source);
+	GsInstalledPage *self = GS_INSTALLED_PAGE (user_data);
 	GtkWidget *widget;
-	guint i;
-	guint cnt = 0;
-	g_autoptr(GsAppList) pending = NULL;
+	guint pending_apps_count = 0;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GsAppList) list = gs_plugin_loader_job_process_finish (plugin_loader,
+									 res,
+									 &error);
+	if (list == NULL) {
+		if (!g_error_matches (error, GS_PLUGIN_ERROR, GS_PLUGIN_ERROR_CANCELLED))
+			g_warning ("failed to refine pending apps: %s", error->message);
+		goto out;
+	}
 
-	/* add new apps to the list */
-	pending = gs_plugin_loader_get_pending (plugin_loader);
-	for (i = 0; i < gs_app_list_length (pending); i++) {
-		app = gs_app_list_index (pending, i);
+	for (guint i = 0; i < gs_app_list_length (list); ++i) {
+		GsApp *app = gs_app_list_index (list, i);
+		if (gs_app_is_installed (app))
+			continue;
 
 		/* never show OS upgrades, we handle the scheduling and
 		 * cancellation in GsUpgradeBanner */
 		if (gs_app_get_kind (app) == AS_APP_KIND_OS_UPGRADE)
 			continue;
 
-		/* do not to add pending apps more than once. */
-		if (gs_installed_page_has_app (self, app) == FALSE)
-			gs_installed_page_add_app (self, pending, app);
+		if (gs_app_get_state (app) == AS_APP_STATE_AVAILABLE)
+			gs_app_set_state (app, AS_APP_STATE_QUEUED_FOR_INSTALL);
 
-		/* incremement the label */
-		cnt++;
+		++pending_apps_count;
```

Comment:
Again, for consistency, I prefer `pending_apps_count++;`, but I can live with it as is if you feel strongly.